metadata-service/auth-filter/src/main/java/com/datahub/auth/authentication/filter/AuthenticationFilter.java

```diff
@@ -0,0 +1,250 @@
+          authenticatorContext);
+      authenticatorChain.register(noOpAuthenticator);
+    }
+  }
+
+  private AuthenticationRequest buildAuthContext(HttpServletRequest request) {
+    return new AuthenticationRequest(Collections.list(request.getHeaderNames())
+        .stream()
+        .collect(Collectors.toMap(headerName -> headerName, request::getHeader)));
+  }
+
+  private void registerPlugins(AuthenticatorChain authenticatorChain) {
+    // TODO: Introduce plugin factory to reduce duplicate code around authentication and authorization processing
+
+    ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+    Path pluginBaseDirectory = Paths.get(configurationProvider.getDatahub().getPlugin().getAuth().getPath());
+    Optional<Config> optionalConfig = (new ConfigProvider(pluginBaseDirectory)).load();
+    optionalConfig.ifPresent((config) -> {
+      log.info("Processing authenticator plugin from auth plugin directory {}", pluginBaseDirectory);
+      PluginConfigFactory authenticatorPluginPluginConfigFactory =
+          new PluginConfigFactory(config);
+
+      List<PluginConfig> authorizers =
+          authenticatorPluginPluginConfigFactory.loadPluginConfigs(PluginType.AUTHENTICATOR);
+      // Filter enabled authenticator plugins
+      List<PluginConfig> enabledAuthenticators = authorizers.stream().filter(pluginConfig -> {
+        if (!pluginConfig.getEnabled()) {
+          log.info(String.format("Authenticator %s is not enabled", pluginConfig.getName()));
+        }
+        return pluginConfig.getEnabled();
+      }).collect(Collectors.toList());
+
+      SecurityMode securityMode =
+          SecurityMode.valueOf(this.configurationProvider.getDatahub().getPlugin().getPluginSecurityMode());
+      // Create permission manager with security mode
+      PluginPermissionManager permissionManager = new PluginPermissionManagerImpl(securityMode);
+
+      // Initiate Authenticators
+      enabledAuthenticators.forEach((pluginConfig) -> {
+        IsolatedClassLoader isolatedClassLoader = new IsolatedClassLoader(permissionManager, pluginConfig);
+        // Create context
+        AuthenticatorContext context = new AuthenticatorContext(
+            ImmutableMap.of(PluginConstant.PLUGIN_HOME, pluginConfig.getPluginHomeDirectory().toString()));
+
+        try {
+          Thread.currentThread().setContextClassLoader((ClassLoader) isolatedClassLoader);
+          Authenticator authenticator = (Authenticator) isolatedClassLoader.instantiatePlugin(Authenticator.class);
+          log.info("Initializing plugin {}", pluginConfig.getName());
+          authenticator.init(pluginConfig.getConfigs().orElse(Collections.emptyMap()), context);
+          authenticatorChain.register(authenticator);
+          log.info("Plugin {} is initialized", pluginConfig.getName());
+        } catch (ClassNotFoundException e) {
+          throw new RuntimeException(e);
```

Comment:
Not useful to throw a wrapped runtime exception without adding a message - please add a message

Reply:
Added message